File: Sep/Day15/maximum-number-of-words-you-can-type.py
class Solution:
    #TC - O(n*b), SC - O(1)
    def canBeTypedWords(self, text: str, brokenLetters: str) -> int:
        count = 0
        for tex in text.split():
            count+=1
            for x in brokenLetters:
                if x in tex:
                    count -=1
                    break
        return count

class Solution:
    #TC - O(n+b), SC - O(1) - beats 100% runtime
    def canBeTypedWords(self, text: str, brokenLetters: str) -> int:
        brokenLetters = set(brokenLetters) #O(b) to convert to a set
        count = 0
        for tex in text.split():
            if not any(ch in brokenLetters for ch in tex):
                count+=1
        return count

class Solution:
    #TC - O(n), SC - O(1) - beats 36.59%
    def canBeTypedWords(self, text: str, brokenLetters: str) -> int:
        bokenLettersArr = [0]*26
        for ch in brokenLetters:
            bokenLettersArr[ord(ch) - 97]=1
        
        wordIsBroken = False
        count = 1
        for ch in text:
            if ch == ' ':
                count+=1
                wordIsBroken = False
            else :
                if not wordIsBroken and bokenLettersArr[ord(ch) - 97] :
                    wordIsBroken = True
                    count -=1
        return count

File: Sep/Day15/maximum-number-of-words-you-can-type.cc
//  Time: O(N * B)  ::  Space:  O(1)
class Solution {
public:
    int canBeTypedWords(string text, string brokenLetters) {
        int count = 0;
        string word;
        stringstream ss(text);
        while (ss >> word) {
            count++;
            for (char x : brokenLetters) {
                if (word.find(x) != string::npos) {
                    count--;
                    break;
                }
            }
        }
        return count;
    }
};



//  Time: O(N + B)  ::  Space:  O(1)
class Solution {
public:
    int canBeTypedWords(string text, string brokenLetters) {
        unordered_set<char> broken(brokenLetters.begin(), brokenLetters.end());
        int count = 0;
        string word;
        stringstream ss(text);
        while (ss >> word) {
            bool isBroken = false;
            for (char ch : word) {
                if (broken.count(ch)) {
                    isBroken = true;
                    break;
                }
            }
            if (!isBroken) count++;
        }
        return count;
    }
};



//  Time: O(N)  ::  Space:  O(1)
class Solution {
public:
    int canBeTypedWords(string text, string brokenLetters) {
        vector<int> brokenArr(26, 0);
        for (char ch : brokenLetters) {
            brokenArr[ch - 'a'] = 1;
        }
        
        bool wordIsBroken = false;
        int count = 1;
        for (char ch : text) {
            if (ch == ' ') {
                count++;
                wordIsBroken = false;
            } else {
                if (!wordIsBroken && brokenArr[ch - 'a']) {
                    wordIsBroken = true;
                    count--;
                }
            }
        }
        return count;
    }
};
